### PR TITLE
Introduces app support for gitcoin grants

### LINKED ISSUE #692: Create Gitcoin Grant report for grant owners
## Feature Description

**Possibly a Premium Only Feature**

Running a Gitcoin grant is cool and if the community values your project and in combination with CLR funding a good amount of funds can flow into an Ethereum account.

It would be really nice if there was a way for users to input an Ethereum account and have it auto-detected that it's a receiving address for a Gitcoin grant.
Once that is done all ERC20 transactions to that grant, along with the USD value of the token at the time can be counted in as profit and reported in a special Grants Tab.

This way later when the owner of the grant uses those tokens elsewhere (say in an exchange) Rotki will know where they came from. Also if the owner needs to report them as income he can use the Rotki grants tab to export all of his inbound transactions into a CSV which can then be imported into a google spreadsheet and share it with their accountant.

### Frontend

- If not yet existing add a left-sidebar top level menu called "Decentralized Finance". Under it should be a sub-menu entry called "Gitcoin Grants".
- This should show a page which will render only if the user has premium. If not it will ask the user to purchase a premium subscription.
- The aforementioned page would show all ethereum accounts that have been detected as grant recipients. For each account it will allow the placing of a label, so that the user can name his grant. (Could we see that from onchain data?). Each account would show a list of all transactions that are grant related, token transferred and USD/EUR/XXX value at the time of transfer. At the end there should be a total of funds received from the grant.
- There should be a date range component which would allow narrowing down the range of the transactions to display.
- There should also be an option with a button to export these data in CSV format.

### Backend

- Only if the user has premium do all the tasks below
- Query all user addresses to see if they own a grant. And then use the gitcoin api endpoint we asked for [here](https://github.com/gitcoinco/web/issues/6493) to get all payment details.
- Add rotki api endpoint to query the data
- Include it in the PnL report

### PR DESCRIPTION
**Pull request by kelsos**
Closes #692

### CODE REVIEW

**Comment by kelsos** [comment]
Check the PR it self and I will see if I can handle the dependency part in the PR that will fix the macos build

**Review by LefterisJP — COMMENTED** [review]
Will leave some UI comments here.

## Color of gitcoin icon

For consistency sake I suppose the gitcoin icon should have no color or be colorized same as the theme?



## Position in history

Putting it in a separate section in history won't scale.



We don't have enough space to do that. In reality gitcoin events are just ledger actions with extra data. This is how they are also saved in the DB.

Perhaps just keep them in the left menu there for now until we figure out a better place to put them? But definitely don't put another parent tab in the history component.

Open to other suggestions. Perhaps we can also have a call on that?

## Gitcoin events table default sort change

Change it to be descending time. So most recent first.



## Show an icon for type

I think it would be a bit cooler and save us a tad bit of space in the table if we showed the type of transaction as an icon. For ethereum show the ethereum logo and for zksync the zksync logo (the 2 arrows thing).

## Different table of events per grant

Please have a different table of events per grant too. Or ... at least somehow show which grants we have information about. Right now there is now way to figure this out. This will also make it much easier for user to understand how to delete a grant's events as for each grant we have data for you can have the trash icon next to it, instead of mixing it in the same place as adding events and fetching report.

## Fetch events for a period is counter-intuitive




I saw the components for the pnl report and assumed they were tied to the gitcoin report. So I had no idea how to specify the period for which to get gitcoin events and as a user you would assume that all gitcoin events until now are retrieved (just like in other places in the app).

Can you separate the menus for report generation and gitcoin event fetching? It's really confusing right now.

Also do you think we should fetch all gitcoin events by default? (May be slow or even timeout if gitcoin's servers have problems).

## No feedback at pressing report button

The report button appears pressed all the time and gives no feedback. User sees nothing happen.


https://user-images.githubusercontent.com/1658405/124966087-225e1380-e023-11eb-9a72-900c1735ff5e.mp4

After a lot of time going through the api messages, I realized something happened and I scrolled down to see the report. Definitely counter-intuitive. I think the separation of the events fetching/showing and the report in different sections of the frontend will help a lot with making this clearer.

## Report period component looks out of place

There is no need to have a separate component just to show a period. 




Wouldn't it be easier and shorter to have the range as part of the title of the report?

`REPORT FOR GITCOIN GRANT 149` -> 

```
REPORT FOR GITCOIN GRANT 149
from 01/07/2020 00:00 to 20/09/2020 23:59:59
```

**Comment by kelsos** [comment]
> ## Position in history
> 
> Putting it in a separate section in history won't scale.
> 
> 
> 
> We don't have enough space to do that. In reality gitcoin events are just ledger actions with extra data. This is how they are also saved in the DB.
> 
> Perhaps just keep them in the left menu there for now until we figure out a better place to put them? But definitely don't put another parent tab in the history component.
> 
> Open to other suggestions. Perhaps we can also have a call on that?

Keeping it in the left menu but not adding a tab is not possible technically. If there is not a tab there is no tab content so even if you have an entry in the left menu it won't render anything. I will see if I can rewrite the component so we can have invisible tabs, but I am not sure if this is feasible.  If you have any suggestions I am open.

In general, though I am not happy with the tabs and how the work in general and I have been thinking, looking for a better alternative, without much luck so far

**Comment by kelsos** [comment]
> Will leave some UI comments here.
> ## Color of gitcoin icon
> 
> For consistency sake I suppose the gitcoin icon should have no color or be colorized same as the theme?
> 

I am not sure how easy would it be to replicate the same exact effect as the font-icons with the png, in the staking we also don't apply any filters in the icons. I could apply a grayscale filter on top but the result will still not be consistent.

**Comment by kelsos** [comment]
> ## No feedback at pressing report button
> 
> The report button appears pressed all the time and gives no feedback. User sees nothing happen.
> Peek.2021-07-08.19-30.mp4
> 
> After a lot of time going through the api messages, I realized something happened and I scrolled down to see the report. Definitely counter-intuitive. I think the separation of the events fetching/showing and the report in different sections of the frontend will help a lot with making this clearer.

I am not sure what you mean pressed all the time, I don't see a pressed button. When pressed the button has a grey area around it.

https://user-images.githubusercontent.com/2269732/125044585-62ada800-e09c-11eb-8d20-406b0442050a.mp4

So the feedback here going away fast that is true, this is for the case that there are no data to report. 

For actual reports, you can see the loading working for a longer period. In the case of actual report data existing, I would assume that seeing a loading complete would be sufficient. 

The no-report data case I can agree that has insufficient feedback, and that is easy to miss and that is especially true, if you are running in software rendering mode where the animations are really slow and frames are skipped.

**Review by LefterisJP — COMMENTED** [review]
Kelsos another thing, since this closes the issue can you also add a changelog entry for what we add? Namely display all gitcoin events per grant and creation of an earnings report per grant for a given period of time

**Review by LefterisJP — COMMENTED** [review]
All right so it's much improved. Some more comments and problems I ran into.

## Make the asset/amount table like the other ones. 

This is now combined in one column and it looks off.

Here is how it looks in the events component for gitcoin grants:


And here is how it looks in other tables. Say ledger actions.




Make it look consistent to the other tables.

## Report generation grant id selector

Thereport generation grant id selector should only be for grants whose events you have already fetched events. So the selector should be a drop-down.



## Assets in the report table



Basically same comment as the events table. Make it looks like the other tables. Where asset contains icon and name/symbol and amount just the amount.

## Fetching events for new grant is not reflected in the UI




I had events for 2 grants with which I was playing. I tried to get events for a new grant. The api responded, but it's nowhere to be seen in the UI. No new section was created.

**Review by LefterisJP — COMMENTED** [review]
I just noticed a python exception when I closed the app I was testing. Not sure how it came to be. Will need to look it up as that's backend.

```
2021-07-09T17:50:00.949Z: Traceback (most recent call last):
  File "src/gevent/greenlet.py", line 906, in gevent._gevent_cgreenlet.Greenlet.run
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/api/rest.py", line 274, in _do_query_async
    result = getattr(self, command)(**kwargs)
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/api/rest.py", line 3150, in _get_gitcoin_events
    only_cache=only_cache,
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/ethereum/gitcoin/api.py", line 194, in query_grant_history
    to_timestamp=period_range[1],
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/ethereum/gitcoin/api.py", line 271, in query_grant_history_period
    action = _deserialize_transaction(grant_id=grant_id, rawtx=transaction)
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/ethereum/gitcoin/api.py", line 69, in _deserialize_transaction
    asset = get_gitcoin_asset(symbol=rawtx['asset'], token_address=rawtx['token_address'])
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/ethereum/gitcoin/api.py", line 46, in get_gitcoin_asset
    return EthereumToken(to_checksum_address(token_address))
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/eth_utils/address.py", line 110, in to_checksum_address
    norm_address = to_normalized_address(value)
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/eth_utils/address.py", line 64, in to_normalized_address
    "Unknown format {}, attempted to normalize to {}".format(value, hex_address)
ValueError: Unknown format 0x0, attempted to normalize to 0x0
2021-07-09T17:45:54Z <Greenlet at 0x7fbdb36fd830: <bound method RestAPI._do_query_async of <rotkehlchen.api.rest.RestAPI object at 0x7fbdb5b51f90>>('_get_gitcoin_events', 12, from_timestamp=1609455600, to_timestamp=1614207600, grant_id=490, only_cache=False)> failed with ValueError
2021-07-09T17:50:00.949Z: rotki REST API server is running at: 127.0.0.1:4242
```

**Comment by LefterisJP** [comment]
So that exception happend when I tried to query the extra grant. Which is why I guess we did not see anything in the UI. But this error should have been shown somewhere.

**Comment by LefterisJP** [comment]
This should fix that exception: https://github.com/rotki/rotki/pull/3170
Also reported a bug in gitcoin:  https://github.com/gitcoinco/web/issues/9280

**Comment by kelsos** [comment]
> All right so it's much improved. Some more comments and problems I ran into.
> ## Make the asset/amount table like the other ones.
> 
> This is now combined in one column and it looks off.
> 
> Here is how it looks in the events component for gitcoin grants:
> 
> 
> And here is how it looks in other tables. Say ledger actions.
> 
> 
> 
> Make it look consistent to the other tables.

I am not sure what looks off other than the header not being properly aligned. Like I mentioned before it was combined before, you just didn't notice :). Also, ledger actions are the wrong example here. In Ledger actions from what I can see, there is no value column so we are having the separate column approach. 

> Make it look consistent to the other tables.
But it is consistent with other tables.  This is the same component that is used all over the Defi section wherever we show a single column of amount/value/symbol. 

See for example Deposits/Liabilities, Yearn Vaults etc.

**Comment by kelsos** [comment]
> I had events for 2 grants with which I was playing. I tried to get events for a new grant. The api responded, but it's nowhere to be seen in the UI. No new section was created.

Did you confirm that any of the responses for the last two async tasks had any events for the specific event id?

The refreshing mechanism does two requests, one to import the data and one to fetch again all the events from the database, so unless you verified the existence of the second grant's id in the result, this means nothing at all and it could very well be that the secondary fetch didn't have any data.

I also did a check to verify, and the only time I managed to replicate what you are describing is when there where no data for the specified grant in the selected period.

When I checked a grant that actually had donations during the period I got a second table appearing properly.

**Review by LefterisJP — APPROVED** [review]
Any new changes should come in the name/created_on addition PR.
